### PR TITLE
Feature generar alertas 159956421

### PR DESCRIPTION
**Pull request by franbe7**
Alert, Alert-User and Alert-Group models
Validations
Controller to create an alert

Alert-User is a relation that represents an alert with an associated user
Alert-Group is analogous to Alert-User

### CODE REVIEW

**Comment by franbe7** [comment]
Using polymorphic associations, the difference is that we will have a relationship model instead of two models (user-alert and alert-group).
I'm right?

**Comment by LeticiaErrandonea** [comment]
@franbe7 Yes, you'll only need one model for the association. Some people also like having a model `Alertable` and make `User` and `Group` to behave like it (see [this article](https://medium.com/@utkukaynar/polymorphic-has-many-through-associations-in-ruby-on-rails-97778e1cea1a) for more info on this approach). I recommend taking a look at both ways of doing it trying to think about upcoming features so then you don't have to change this too much. 

Disclaimer: I haven't checked the upcoming features yet. I'm just bringing this options up but don't have a strong opinion towards either of them

**Comment by franbe7** [comment]
@LeticiaErrandonea We decided for the first aproach in the link.
It is already implemented and we did the rebase for the login feature and change the tests

File: app/controllers/alerts_controller.rb [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      result = []
+      result.push(alert)
+      result = add_users(result, alert.id)\
+       if params[:users].present?
+      result = add_groups(result, alert.id)\
+       if params[:groups].present?
+      result
+    end
+
+    def invalid_operation
```
Comment (LeticiaErrandonea) [inline app/controllers/alerts_controller.rb]: This could be a validation at model level. In Rails we prefer having in the models everything that the model knows how to do (including validations).

That way, in the controller you just try to save it and don't have to take care of the reason of being invalid.

File: app/controllers/alerts_controller.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    elsif @alert.save
+      render json: success(@alert), status: :created
+    else
+      render json: @alert.errors, status: :unprocessable_entity
+    end
+  end
+
+  private
+
+    def success(alert)
```
Comment (LeticiaErrandonea) [inline app/controllers/alerts_controller.rb]: What is this method doing? Is it serializing the created alert?

File: app/controllers/alerts_controller.rb [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+      result = []
+      result.push(alert)
+      result = add_users(result, alert.id)\
```
Comment (LeticiaErrandonea) [inline app/controllers/alerts_controller.rb]: nit: `if` should be in the same line

File: app/controllers/alerts_controller.rb [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+      result = add_groups(result, alert.id)\
+       if params[:groups].present?
+      result
+    end
+
+    def invalid_operation
+      params[:users].blank? && params[:groups].blank? || !@alert.save
+    end
+
+    def add_users(result, alert_id)
```
Comment (LeticiaErrandonea) [inline app/controllers/alerts_controller.rb]: Same as above: you could handle the creation of `alert_user` in a `after_save` callback in `user` and `group`, so you keep the controller smaller.

File: app/models/alert_association.rb [hunk 5]
```diff
@@ -0,0 +1,4 @@
+class AlertAssociation < ApplicationRecord
+  belongs_to :alert, class_name: 'Alert'
```
Comment (LeticiaErrandonea) [inline app/models/alert_association.rb]: nit: I don't think you need to specify the class name here

File: app/models/alert.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  has_many :alert_associations, dependent: :destroy
+  has_many :users, through: :alert_associations, source: :alerted, source_type: 'User'
+  has_many :groups, through: :alert_associations, source: :alerted, source_type: 'Group'
+  enum alert_type: { info: 0, warning: 1, alert: 2 }
+  validates :description, presence: true
+  validates :expire_at, presence: true
+  belongs_to :author, class_name: 'User'
+  validate :expire_at_cannot_be_in_the_past
+
+  def expire_at_cannot_be_in_the_past
```
Comment (LeticiaErrandonea) [inline app/models/alert.rb]: nit: validations should be private methods

File: app/models/alert.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
+class Alert < ApplicationRecord
+  has_many :alert_associations, dependent: :destroy
+  has_many :users, through: :alert_associations, source: :alerted, source_type: 'User'
+  has_many :groups, through: :alert_associations, source: :alerted, source_type: 'Group'
+  enum alert_type: { info: 0, warning: 1, alert: 2 }
+  validates :description, presence: true
+  validates :expire_at, presence: true
+  belongs_to :author, class_name: 'User'
```
Comment (LeticiaErrandonea) [inline app/models/alert.rb]: nit: all associations should be together. 
Also, ideally, you'd have blocks of associations and validations, separated by empty newlines:
```
has_many ...
belongs_to ...
has_one ...

validates ...
validates ...

enum ...
```

File: app/models/alert_association.rb [hunk 8]
```diff
@@ -0,0 +1,4 @@
+class AlertAssociation < ApplicationRecord
```
Comment (LeticiaErrandonea) [inline app/models/alert_association.rb]: nit: the model name should be self-explanatory so by just reading the name you can figure out the models involved. WDYT about `AlertAlerted` or something like that?

File: config/routes.rb [hunk 9]
```diff
@@ -1,5 +1,6 @@
 Rails.application.routes.draw do
-  resources :groups, only: [:create, :show], as: 'group'
+  resources :alerts, only: [:create], as: 'alert'
```
Comment (LeticiaErrandonea) [inline config/routes.rb]: Why are you removing the `s` from all the routes?

File: spec/controllers/alerts_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,71 @@
+require 'rails_helper'
+
+RSpec.describe AlertsController, type: :controller do
+
+  let(:user) { create(:user, :valid) }
+  let(:groups) do
+    groups = []
+    group_hash = {}
+    3.times do
```
Comment (LeticiaErrandonea) [inline spec/controllers/alerts_controller_spec.rb]: nit: you could state that you want 3 users as a param for FactoryBot (check [docs](https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#building-or-creating-multiple-records)).

File: spec/factories/groups.rb [hunk 11]
```diff
@@ -1,17 +1,17 @@
...
+      image { FFaker::Image.url }
+      admin_id { 1 }
     end
 
     trait :invalid do
-      name ""
-      description ""
-      image ""
-      admin_id nil
+      name { "" }
```
Comment (LeticiaErrandonea) [inline spec/factories/groups.rb]: nit: these changes doesn't seem to belong to this PR

File: app/controllers/alerts_controller.rb (hunk 1 above)
Comment (franbe7) [reply app/controllers/alerts_controller.rb]: The problem that I see is that the group and user lists are not alert model attributes.
In this case, is there a way to validate this in the model?

Comment (franbe7) [reply app/controllers/alerts_controller.rb]: The problem that i see is that the group and user lists are not alert model attributes.
In this case, is there a way to validate this in the model?

Comment (LeticiaErrandonea) [reply app/controllers/alerts_controller.rb]: From the top of my head I think that maybe `AlertAssociation` could work but I haven't given it much thought. You'll have to investigate a little bit where is best to place it.

File: app/controllers/alerts_controller.rb (hunk 2 above)
Comment (LeticiaErrandonea) [reply app/controllers/alerts_controller.rb]: As I stated in a different comment, in the controller we shouldn't be creating the intermediate model. Also, aren't there any Rails methods (or Ruby gems) to create a JSON from an object and its associations? I think [`as_json`](https://api.rubyonrails.org/classes/ActiveModel/Serializers/JSON.html#method-i-as_json) should work

File: spec/factories/groups.rb (hunk 11 above)
Comment (franbe7) [reply spec/factories/groups.rb]: Yes. This belong here. Otherwise rspec fails.

File: config/routes.rb (hunk 9 above)
Comment (franbe7) [reply config/routes.rb]: We don't understand this, what do you want say?

File: spec/factories/groups.rb (hunk 11 above)
Comment (LeticiaErrandonea) [reply spec/factories/groups.rb]: But this was working before without `{}`, right? In `master` these `{}` aren't present. Aren't the specs passing in `master`?

File: config/routes.rb (hunk 9 above)
Comment (LeticiaErrandonea) [reply config/routes.rb]: By default, Rails routes are `alerts` and `groups`. You are making them be `alert` and `group` by using the `as` option. Why is it that you are you doing that?

Comment (franbe7) [reply config/routes.rb]: Once we had a problem with routes and we added this, but maybe the error was other. Also we use the plural routes, so maybe it's not necessary.

**Comment by franbe7** [comment]
@LeticiaErrandonea what do you think about the changes requested? only be necessary to change the routes problem?

File: config/routes.rb (hunk 9 above)
Comment (LeticiaErrandonea) [reply config/routes.rb]: From [Rails docs](https://guides.rubyonrails.org/routing.html#overriding-the-named-helpers), with the `as` option you're changing the helpers (`alerts_path` is `alert_path`). I don't think you're using any helpers, so this isn't affecting the behaviour of the API. It'd be good to remove the `as` option in the PRs that are adding it and that haven't been merged, so we don't add unnecessary config. For the future routes that you may need, just keep in mind not to add this option (unless you find out you need it, of course).

File: spec/factories/groups.rb (hunk 11 above)
Comment (LeticiaErrandonea) [reply spec/factories/groups.rb]: 👍cool, thanks for the explanation. Just in case, for the `invalid` one it will always create the groups with the same params. I like that it has `{}`, though. Makes the file consistent.

File: spec/models/alert_spec.rb [hunk 12]
```diff
@@ -0,0 +1,37 @@
+require 'rails_helper'
+
+RSpec.describe Alert, type: :model do
+  describe "Validations" do
+    let(:user) { create(:user, :valid) }
+    let(:user_list) { create_list(:user, 3, :valid) }
+    let(:users) do
+      users = []
+      user_hash = {}
```
Comment (LeticiaErrandonea) [inline spec/models/alert_spec.rb]: nit: you don't need to initialize an empty hash. You can just initialize it with the data: `some_hash = { key1: "value1" }`

Comment (LeticiaErrandonea) [reply spec/models/alert_spec.rb]: Also, why do you need the `user_hash`?

File: config/routes.rb (hunk 9 above)
Comment (franbe7) [reply config/routes.rb]: Done
